### PR TITLE
refactor: standalone spec parsing for parameters

### PR DESCRIPTION
**Pull request by lukasheinrich**
# Description

this simplifies the `_ModelConfig` `__init__` method and encapsulates the code that needs the JSON structure (somewhat related to #742 ) so that it could be replaced maybe by something non-JSON-y

This is a simple refactor.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* refactor spec usage into standalone parsing function
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
@lukasheinrich You had mentioned you were confused about the Codecov report. What parts in particular? The diff shows that all the unhit lines are the `raise RuntimeError` lines.

**Comment by lukasheinrich** [comment]
I'm confused because in the github comment it says

```
Merging #744 into master will increase coverage by <.01%.
```

so I don't understand why it's complaining

**Comment by matthewfeickert** [comment]
> I'm confused because in the github comment it says
> 
> ```
> Merging #744 into master will increase coverage by <.01%.
> ```
> 
> so I don't understand why it's complaining

Yes, but that's the total coverage. The part that is complaining is the **patch** coverage, which is low:

patch_coverage

**Comment by lukasheinrich** [comment]
done

File: src/pyhf/pdf.py [hunk 1]
```diff
@@ -17,90 +17,217 @@
...
+    _reqs = reduce_paramsets_requirements(
+        _paramsets_requirements, _paramsets_user_configs
+    )
+
+    _sets = {}
+    for param_name, paramset_requirements in _reqs.items():
+        paramset_type = paramset_requirements.get('paramset_type')
+        paramset = paramset_type(**paramset_requirements)
+        _sets[param_name] = paramset
+    return _sets
```
Comment (kratsg) [inline src/pyhf/pdf.py]: These lines are specific to a given Model. If we care about efficiency and running things once and passing around that object, then we want to factor this out into a separate function called by the Model, with this existing function called by the Workspace, perhaps renaming this as `_paramset_requirements_from_channels`.

**Pull request merged**